src/boot/bootctl-install.c

```diff
@@ -449,7 +455,156 @@ static int copy_one_file(const char *esp_path, const char *name, bool force) {
+                char** ret_loader_path, char** ret_description)
+{
+        _cleanup_free_ char *buf = NULL, *loader_path = NULL, *description = NULL;
+        const char *s = NULL;
+        int r;
+
+        /* enforce unicode LE to avoid decoding garbage */
+        if (*csv != 0xfeff)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid boot.csv encoding: %hx", *csv);
+
+        buf = utf16_to_utf8(csv+1, size-2);
+        if (!buf)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to decode boot.csv. Must be UTF-16");
+
+        /* only care about first line */
+        s = buf = truncate_nl(buf);
+
+        r = extract_first_word(&s, &loader_path, ",", EXTRACT_KEEP_QUOTE);
+         /* shortest file name x.efi has four chars */
+        if (r < 0 || strlen(loader_path) < 4)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Missing or too short file name in boot.csv");
+
+        r = extract_first_word(&s, &description, ",", EXTRACT_KEEP_QUOTE);
+        if (r < 0 || !strlen(description))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Missing description in boot.csv");
+
+        if (ret_loader_path)
+                *ret_loader_path = TAKE_PTR(loader_path);
+
+        if (ret_description)
+                *ret_description = TAKE_PTR(description);
+
+        return 0;
+}
+
+struct CopyLoaderArgs {
+        const char* source_path;
+        const char* esp_path;
+        bool force;
+        char** loader_path;
+        char** loader_description;
+};
+
+static int copy_loader(
+                RecurseDirEvent event,
+                const char *path,
+                int dir_fd,
+                int inode_fd,
+                const struct dirent *de,
+                const struct statx *sx,
+                void *userdata) {
+
+        _cleanup_free_ char *source_path = NULL, *dest_path = NULL;
+        struct CopyLoaderArgs *args = userdata;
+        int r;
+
+        assert(path);
+        assert(args);
+        assert(args->source_path);
+        assert(args->esp_path);
+
+        if (event != RECURSE_DIR_ENTRY)
+                return RECURSE_DIR_CONTINUE;
+
+        r = chase(path, args->esp_path, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS|CHASE_NONEXISTENT, &dest_path, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to resolve path %s under directory %s: %m", path, args->esp_path);
+
+        source_path = path_join(args->source_path, path);
+
+        r = copy_file_with_version_check(source_path, dest_path,
+                COPY_WITH_VERSION_CHECK_IGNORE_MISSING_VERSION | (args->force ? COPY_WITH_VERSION_CHECK_FORCE : 0));
+
+        /* parse only the first boot.csv we find if requested */
+        if (!args->loader_path || !args->loader_description
+                        || *args->loader_path || *args->loader_description
+                        || !endswith_no_case(path, "boot.csv") || (r < 0 && !IN_SET(r, -ESTALE, -ESRCH)))
```

Comment:
if you are looking for the file name without a prefixed full path, use `de->d_name`